src/ripple/basics/impl/FileUtilities.cpp

```diff
@@ -0,0 +1,63 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/FileUtilities.h>
+
+namespace ripple
+{
+
+std::string getFileContents(boost::system::error_code& ec,
+    boost::filesystem::path const& sourcePath,
+    boost::optional<std::size_t> maxSize)
+{
+    using namespace boost::filesystem;
+    using namespace boost::system::errc;
+
+    path fullPath{ canonical(sourcePath, ec) };
+    if (ec)
+        return {};
+
+    if (maxSize && (file_size(fullPath, ec) > *maxSize || ec))
+    {
+        if (!ec)
+            ec = make_error_code(file_too_large);
+        return {};
+    }
+
+    ifstream fileStream(fullPath, std::ios::in);
+
+    if (!fileStream)
+    {
+        ec = make_error_code(static_cast<errc_t>(errno));
```

Comment:
Ugh... really? Is this portable on Windows and MacOS?

I dug around, trying to figure out how to get `std::ifstream` to throw reasonable exceptions or return `std::error_code`-based error codes. The closest I came was (`std::basic_ios::exceptions`)(https://en.cppreference.com/w/cpp/io/basic_ios/exceptions) and [`std::ios_base::failure`](https://en.cppreference.com/w/cpp/io/ios_base/failure) and, honestly, having stared at the docs, I have no idea how to use the former to generate the latter.

Reply:
As far as I can tell, yes.  The `errc_t` enum is defined in terms of the `E...` codes that are `#define`d in errno.h. (boost/boost/system/error_code.hpp)

Reply:
Although this works and is likely based on boost, you might want to maintain the library and type consistency by using `boost::system::errc::make_error_code`.

Reply:
@miguelportilla I'm pretty sure I'm already getting the type consistency, because all the overrides of `make_error_code` that I could find have strongly types parameters, and since I'm passing a typed parameter, it'll always get the right override. Also I have an appropriate `using namespace` at the top of the function.

Reply:
@ximinez I could have sworn I read `std::make_error_code`. My tired eyes deceive me, time for a walk.

Reply:
No worries! Enjoy your walk. :smile: